File: moransI.R
d <- read.table("/Users/singhal/thesisWork/introgression/LD/moransI.out",header=F)
names(d) <- c("contact","dist","moransI")

contacts <- c("sjo","nBMB","carlia", "gillies")
colors <- c("#F8766D","#7CAE00","#00BFC4","#C77CFF")

d[d$dist == 0,]$dist = 1.1
plot(NULL,xlim=range(d$dist),ylim=range(d$moransI),xlab="distance (bps)", ylab="Moran's I")

for (i in 1:length(contacts)) {
	tmp <- d[d$contact==contacts[i],]
	lines(tmp$dist,tmp$moransI,col=colors[i],lwd=2)	
}
legend(x=700,y=0.55,legend=contacts,fill=colors,col=colors,bty="n")

File: moransI.pl
use warnings;
use strict;

my @contact = qw(sjo nBMB gillies carlia);
#my @contact = qw(carlia);
my @dist = qw(0 100 200 300 400 500 600 700 800 900 1000);

foreach my $contact (@contact) {
	my $file = "/Users/singhal/thesisWork/introgression/LD/LD_" . $contact . ".out";

	my %d;
	open(IN,"<$file");
	while(<IN>) {
		chomp(my $line = $_);
		
		my @d = split(/\t/,$line);
		push(@{$d{$d[2]}},\@d); 
	
		}
	close(IN);	

	for (my $n = 0; $n < scalar(@dist) - 1; $n++) {
	
		my $nLoci = 0;
		my $sumNumerator = 0;
		my $nComparisons = 0;
		my $sumDenominator = 0;	
	
		my $dist1 = $dist[$n];
		my $dist2 = $dist[$n + 1];
	
		foreach my $c (keys %d) {
			my @tmp = @{$d{$c}};
			if (scalar(@tmp) > 1) {
				$nLoci += scalar(@tmp);  
			
				for (my $i = 0; $i < scalar(@tmp); $i++) {
					$sumDenominator += $tmp[$i][4] ** 2;
					}
			
				for (my $j = 0; $j < scalar(@tmp); $j++) {
					for (my $k = $j + 1; $k < scalar(@tmp); $k++) {
						if ( abs($tmp[$j][3] - $tmp[$k][3]) >= $dist1) {
							if ( abs($tmp[$j][3] - $tmp[$k][3]) <= $dist2) {
								$nComparisons++;
								$sumNumerator += ($tmp[$j][4] * $tmp[$k][4]);
								}
							}	
						}			
					}	
				}
			}	
		my $moranI = ($nLoci * $sumNumerator) / ($nComparisons * $sumDenominator);
		print $contact, "\t", $dist1, "\t", $moranI, "\t$nComparisons\n";
		}
	}	
	